File: docx_parsers/numbering/numbering_parser.py
from typing import List, Optional
from lxml import etree
from docx_parsers.utils import extract_xml_root_from_docx, read_binary_from_file_path, convert_twips_to_points
from docx_parsers.helpers.common_helpers import extract_element, extract_attribute, NAMESPACE
from docx_parsers.models.numbering_models import NumberingLevel, NumberingInstance, NumberingSchema
from docx_parsers.models.styles_models import FontProperties, IndentationProperties
from docx_parsers.styles.paragraph_properties_parser import ParagraphPropertiesParser
import json

class NumberingParser:
    """
    Parses the numbering definitions from a DOCX file.

    This class extracts and parses the numbering definitions found in the 
    numbering.xml file of a DOCX document, converting them into structured 
    Pydantic models for further processing or conversion to other formats.
    """

    def __init__(self, docx_file: bytes):
        """
        Initializes the NumberingParser with the given DOCX file.

        Args:
            docx_file (bytes): The binary content of the DOCX file.
        """
        self.root = extract_xml_root_from_docx(docx_file, 'numbering.xml')
        self.numbering_schema = self.parse()

    def parse(self) -> NumberingSchema:
        """
        Parses the numbering XML into a NumberingSchema.

        Returns:
            NumberingSchema: The parsed numbering schema.

        Example:
            The following is an example of a numbering definition in a numbering.xml file:

            .. code-block:: xml

                <w:numbering>
                    <w:num w:numId="1">
                        <w:abstractNumId w:val="0"/>
                    </w:num>
                    <w:abstractNum w:abstractNumId="0">
                        <w:lvl w:ilvl="0">
                            <w:start w:val="1"/>
                            <w:numFmt w:val="decimal"/>
                            <w:lvlText w:val="%1."/>
                            <w:lvlJc w:val="left"/>
                        </w:lvl>
                    </w:abstractNum>
                </w:numbering>
        """
        instances = []
        for num in self.root.findall(".//w:num", namespaces=NAMESPACE):
            numId = int(extract_attribute(num, 'numId'))
            abstractNumId = int(extract_attribute(extract_element(num, ".//w:abstractNumId"), 'val'))
            levels = self.extract_levels(abstractNumId)
            instance = NumberingInstance(numId=numId, levels=levels)
            instances.append(instance)
        return NumberingSchema(instances=instances)

    def extract_levels(self, abstractNumId: int) -> List[NumberingLevel]:
        """
        Extracts the levels for a given abstract numbering ID.

        Args:
            abstractNumId (int): The abstract numbering ID.

        Returns:
            List[NumberingLevel]: The list of extracted numbering levels.
        """
        levels = []
        abstractNum = extract_element(self.root, f".//w:abstractNum[@w:abstractNumId='{abstractNumId}']")
        for lvl in abstractNum.findall(".//w:lvl", namespaces=NAMESPACE):
            level = self.extract_level(abstractNumId, lvl)
            levels.append(level)
        return levels

    def extract_level(self, numId: int, lvl: etree.Element) -> NumberingLevel:
        """
        Extracts a single numbering level.

        Args:
            numId (int): The numbering ID.
            lvl (etree.Element): The XML element representing the numbering level.

        Returns:
            NumberingLevel: The extracted numbering level.

        Example:
            The following is an example of a numbering level in a numbering.xml file:

            .. code-block:: xml

                <w:lvl w:ilvl="0">
                    <w:start w:val="1"/>
                    <w:numFmt w:val="decimal"/>
                    <w:lvlText w:val="%1."/>
                    <w:lvlJc w:val="left"/>
                </w:lvl>
        """
        ilvl = int(extract_attribute(lvl, 'ilvl'))
        start = int(extract_attribute(extract_element(lvl, ".//w:start"), 'val'))
        numFmt = extract_attribute(extract_element(lvl, ".//w:numFmt"), 'val')
        lvlText = extract_attribute(extract_element(lvl, ".//w:lvlText"), 'val')
        lvlJc = extract_attribute(extract_element(lvl, ".//w:lvlJc"), 'val')

        indent_properties = self.extract_indentation(lvl)
        tab_pt = self.extract_tab(lvl)
        fonts = self.extract_fonts(lvl)

        return NumberingLevel(
            numId=numId, ilvl=ilvl, start=start, numFmt=numFmt, lvlText=lvlText, lvlJc=lvlJc,
            indent=indent_properties, tab_pt=tab_pt, fonts=fonts
        )

    def extract_indentation(self, lvl: etree.Element) -> Optional[IndentationProperties]:
        """
        Extracts indentation properties from a numbering level.

        Args:
            lvl (etree.Element): The XML element representing the numbering level.

        Returns:
            Optional[IndentationProperties]: The extracted indentation properties.

        Example:
            The following is an example of paragraph properties with indentation in a numbering level:

            .. code-block:: xml

                <w:pPr>
                    <w:ind w:left="720" w:hanging="360"/>
                </w:pPr>
        """
        pPr = extract_element(lvl, ".//w:pPr")
        if pPr is not None:
            return ParagraphPropertiesParser().extract_indentation(pPr)

    def extract_tab(self, lvl: etree.Element) -> Optional[float]:
        """
        Extracts tab stop properties from a numbering level.

        Args:
            lvl (etree.Element): The XML element representing the numbering level.

        Returns:
            Optional[float]: The tab stop position in points.

        Example:
            The following is an example of paragraph properties with a tab stop in a numbering level:

            .. code-block:: xml

                <w:pPr>
                    <w:tabs>
                        <w:tab w:val="left" w:pos="720"/>
                    </w:tabs>
                </w:pPr>
        """
        pPr = extract_element(lvl, ".//w:pPr")
        if pPr is not None:
            tab_element = extract_element(pPr, ".//w:tab")
            if tab_element is not None:
                tab_val = extract_attribute(tab_element, 'pos')
                if tab_val and tab_val.isdigit():
                    return convert_twips_to_points(int(tab_val))
        return None

    def extract_fonts(self, lvl: etree.Element) -> Optional[FontProperties]:
        """
        Extracts font properties from a numbering level.

        Args:
            lvl (etree.Element): The XML element representing the numbering level.

        Returns:
            Optional[FontProperties]: The extracted font properties.

        Example:
            The following is an example of run properties with font settings in a numbering level:

            .. code-block:: xml

                <w:rPr>
                    <w:rFonts w:ascii="Calibri" w:hAnsi="Calibri"/>
                </w:rPr>
        """
        rPr = extract_element(lvl, ".//w:rPr")
        if rPr is not None:
            rFonts = extract_element(rPr, "w:rFonts")
            if rFonts is not None:
                return FontProperties(
                    ascii=extract_attribute(rFonts, 'ascii'),
                    hAnsi=extract_attribute(rFonts, 'hAnsi'),
                    eastAsia=extract_attribute(rFonts, 'eastAsia'),
                    cs=extract_attribute(rFonts, 'cs')
                )
        return None

    def get_numbering_schema(self) -> NumberingSchema:
        """
        Gets the parsed numbering schema.

        Returns:
            NumberingSchema: The parsed numbering schema.
        """
        return self.numbering_schema

if __name__ == "__main__":
    docx_path = "C:/Users/omerh/Desktop/new_docx.docx"
    docx_file = read_binary_from_file_path(docx_path)

    numbering_parser = NumberingParser(docx_file)
    filtered_schema_dict = numbering_parser.numbering_schema.model_dump(exclude_none=True)
    print(json.dumps(filtered_schema_dict, indent=2))

File: docx_parser_converter_ts/src/docx_parsers/numbering/numberingParser.ts
import type { NumberingLevel, NumberingInstance, NumberingSchema } from '../models/numberingModels';
import type { FontProperties, IndentationProperties } from '../models/stylesModels';
import { extractElement, extractAttribute, NAMESPACE } from '../helpers/commonHelpers';
import { extractXmlRootFromDocx, convertTwipsToPoints, extractXmlRootFromString } from '../utils';
import { ParagraphPropertiesParser } from '../styles/paragraphPropertiesParser';

/**
 * Parses the numbering definitions from a DOCX file.
 *
 * This class extracts and parses the numbering definitions found in the 
 * numbering.xml file of a DOCX document, converting them into structured 
 * TypeScript interfaces for further processing or conversion to other formats.
 */
export class NumberingParser {
    private root: Element;
    private numberingSchema: NumberingSchema;

    /**
     * Creates a new instance of NumberingParser.
     * 
     * @param source - Either the binary content of the DOCX file
     *                 or the numbering.xml content as a string.
     * @returns A Promise that resolves to a new NumberingParser instance.
     */
    public static async create(source: Uint8Array | ArrayBuffer | string | Element): Promise<NumberingParser> {
        const parser = new NumberingParser();
        await parser.initialize(source);
        return parser;
    }

    private constructor() {
        // Private constructor to enforce using the static create method
        this.root = {} as Element;
        this.numberingSchema = { instances: [] };
    }

    /**
     * Initializes the parser with the given source.
     * 
     * @param source - Either the binary content of the DOCX file,
     *                 the numbering.xml content as a string,
     *                 or a pre-parsed XML Element.
     */
    private async initialize(source: Uint8Array | ArrayBuffer | string | Element): Promise<void> {
        if (source instanceof Element) {
            this.root = source;
        } else if (typeof source === 'string') {
            this.root = extractXmlRootFromString(source);
        } else if (source instanceof Uint8Array || source instanceof ArrayBuffer) {
            this.root = await extractXmlRootFromDocx(source, 'numbering.xml');
        } else {
            // This case should ideally not be reached if type hints are respected.
            // If it is, it means 'source' is not an Element, string, Uint8Array, or ArrayBuffer.
            throw new Error('Invalid source type provided to NumberingParser. Expected Element, string, Uint8Array, or ArrayBuffer.');
        }
        this.numberingSchema = this.parse();
    }

    /**
     * Parses the numbering XML into a NumberingSchema.
     *
     * @returns The parsed numbering schema.
     */
    private parse(): NumberingSchema {
        const instances: NumberingInstance[] = [];
        const nums = this.root.getElementsByTagNameNS(NAMESPACE.w, 'num');

        for (let i = 0; i < nums.length; i++) {
            const num = nums[i];
            const numIdStr = extractAttribute(num, 'numId');
            if (!numIdStr) continue; // Skip if no numId

            const numId = parseInt(numIdStr, 10);
            const abstractNumIdEl = extractElement(num, './/w:abstractNumId');
            const abstractNumIdStr = abstractNumIdEl ? extractAttribute(abstractNumIdEl, 'val') : null;
            if (!abstractNumIdStr) continue; // Skip if no abstractNumId

            const abstractNumId = parseInt(abstractNumIdStr, 10);
            const levels = this.extractLevels(abstractNumId);
            const instance: NumberingInstance = { numId, levels };
            instances.push(instance);
        }

        return { instances };
    }

    /**
     * Extracts the levels for a given abstract numbering ID.
     *
     * @param abstractNumId - The abstract numbering ID.
     * @returns The list of extracted numbering levels.
     */
    private extractLevels(abstractNumId: number): NumberingLevel[] {
        const levels: NumberingLevel[] = [];
        const abstractNum = extractElement(this.root, `//w:abstractNum[@w:abstractNumId='${abstractNumId}']`);

        if (abstractNum) {
            const lvls = abstractNum.getElementsByTagNameNS(NAMESPACE.w, 'lvl');
            for (let i = 0; i < lvls.length; i++) {
                const lvl = lvls[i];
                const level = this.extractLevel(abstractNumId, lvl);
                if (level) {
                    levels.push(level);
                }
            }
        }

        return levels;
    }

    /**
     * Extracts a single numbering level.
     *
     * @param numId - The numbering ID.
     * @param lvl - The XML element representing the numbering level.
     * @returns The extracted numbering level.
     */
    private extractLevel(numId: number, lvl: Element): NumberingLevel | null {
        const ilvlStr = extractAttribute(lvl, 'ilvl');
        if (!ilvlStr) return null;
        const ilvl = parseInt(ilvlStr, 10);

        const startEl = extractElement(lvl, './/w:start');
        const startStr = startEl ? extractAttribute(startEl, 'val') : '1';
        const start = parseInt(startStr || '1', 10);

        const numFmtEl = extractElement(lvl, './/w:numFmt');
        const numFmt = numFmtEl ? extractAttribute(numFmtEl, 'val') : 'decimal';

        const lvlTextEl = extractElement(lvl, './/w:lvlText');
        const lvlText = lvlTextEl ? extractAttribute(lvlTextEl, 'val') : '';

        const lvlJcEl = extractElement(lvl, './/w:lvlJc');
        const lvlJc = lvlJcEl ? extractAttribute(lvlJcEl, 'val') : 'left';

        const indent = this.extractIndentation(lvl);
        const tabPt = this.extractTab(lvl);
        const fonts = this.extractFonts(lvl);

        return {
            numId,
            ilvl,
            start,
            numFmt: numFmt || 'decimal',
            lvlText: lvlText || '',
            lvlJc: lvlJc || 'left',
            ...(indent && { indent }),
            ...(tabPt !== undefined && { tabPt }),
            ...(fonts && { fonts })
        };
    }

    /**
     * Extracts indentation properties from a numbering level.
     *
     * @param lvl - The XML element representing the numbering level.
     * @returns The extracted indentation properties.
     */
    private extractIndentation(lvl: Element): IndentationProperties | undefined {
        const pPr = extractElement(lvl, './/w:pPr');
        if (pPr) {
            const parser = new ParagraphPropertiesParser();
            const props = parser.parse(pPr);
            return props.indent;
        }
        return undefined;
    }

    /**
     * Extracts tab stop properties from a numbering level.
     *
     * @param lvl - The XML element representing the numbering level.
     * @returns The tab stop position in points.
     */
    private extractTab(lvl: Element): number | undefined {
        const pPr = extractElement(lvl, './/w:pPr');
        if (pPr) {
            const tabEl = extractElement(pPr, './/w:tab');
            if (tabEl) {
                const pos = extractAttribute(tabEl, 'pos');
                if (pos && !isNaN(parseInt(pos, 10))) {
                    return convertTwipsToPoints(parseInt(pos, 10));
                }
            }
        }
        return undefined;
    }

    /**
     * Extracts font properties from a numbering level.
     *
     * @param lvl - The XML element representing the numbering level.
     * @returns The extracted font properties.
     */
    private extractFonts(lvl: Element): FontProperties | undefined {
        const rPr = extractElement(lvl, './/w:rPr');
        if (rPr) {
            const rFonts = extractElement(rPr, 'w:rFonts');
            if (rFonts) {
                const ascii = extractAttribute(rFonts, 'ascii') || undefined;
                const hAnsi = extractAttribute(rFonts, 'hAnsi') || undefined;
                const eastAsia = extractAttribute(rFonts, 'eastAsia') || undefined;
                const cs = extractAttribute(rFonts, 'cs') || undefined;
                
                if (ascii || hAnsi || eastAsia || cs) {
                    return { ascii, hAnsi, eastAsia, cs };
                }
            }
        }
        return undefined;
    }

    /**
     * Gets the parsed numbering schema.
     *
     * @returns The parsed numbering schema.
     */
    public getNumberingSchema(): NumberingSchema {
        return this.numberingSchema;
    }
}

